src/tool_operate.c

```diff
@@ -686,19 +686,210 @@ static void single_transfer_cleanup(struct OperationConfig *config)
+        gateway = config->ipfs_gateway;
+    }
+    else {
+      result = CURLE_BAD_FUNCTION_ARGUMENT;
+      goto clean;
+    }
+  }
+  else {
+    gateway = ipfs_gateway();
+    if(!gateway) {
+      result = CURLE_FILE_COULDNT_READ_FILE;
+      goto clean;
+    }
+
+    if(curl_url_set(ipfsurl, CURLUPART_URL, gateway, CURLU_GUESS_SCHEME
+                    | CURLU_NON_SUPPORT_SCHEME) != CURLUE_OK) {
+      goto clean;
+    }
+  }
+
+  pathbuffer = aprintf("%s/%s", protocol, cid);
+  if(!pathbuffer) {
+    goto clean;
+  }
+
+  if(curl_url_set(ipfsurl, CURLUPART_PATH, pathbuffer, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_URL, url, CURLU_URLENCODE)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  if(curl_url_get(ipfsurl, CURLUPART_SCHEME, &protocol, CURLU_DEFAULT_SCHEME)
+                  != CURLUE_OK) {
+    goto clean;
+  }
+
+  result = CURLE_OK;
+
+clean:
+  if(gateway) {
+    curl_free(gateway);
+  }
+  if(cid) {
+    curl_free(cid);
+  }
+  if(pathbuffer) {
+    curl_free(pathbuffer);
+  }
+  if(ipfsurl) {
+    curl_url_cleanup(ipfsurl);
+  }
+
+  return result;
+}
+
 /*
  * Return the protocol token for the scheme used in the given URL
  */
-static const char *url_proto(char *url)
+static const char *url_proto(char **url,
+                             struct OperationConfig *config,
+                             struct GlobalConfig *global)
 {
   CURLU *uh = curl_url();
   const char *proto = NULL;
 
   if(uh) {
-    if(url) {
-      if(!curl_url_set(uh, CURLUPART_URL, url,
-                       CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)) {
+    if(*url) {
+      if(curl_url_set(uh, CURLUPART_URL, *url,
+                      CURLU_GUESS_SCHEME | CURLU_NON_SUPPORT_SCHEME)
+         == CURLUE_OK) {
         char *schemep = NULL;
+
+        if(curl_url_get(uh, CURLUPART_SCHEME, &schemep,
+                         CURLU_DEFAULT_SCHEME) == CURLUE_OK
+           && schemep) {
+          if(curl_strnequal(schemep, "ipfs", 4)
+             || curl_strnequal(schemep, "ipns", 4)) {
+              CURLcode ret = ipfs_url_rewrite(uh, schemep, url, config);
+            if(ret != CURLE_OK) {
+              switch(ret) {
+              case CURLE_URL_MALFORMAT:
+                helpf(global->errors, "malformed URL. Visit https://curl.se/"
+                      "docs/ipfs.html for more information\n");
+                break;
+              case CURLE_FILE_COULDNT_READ_FILE:
+                helpf(global->errors, "IPFS automatic gateway detection fail"
+                      "ure. Visit https://curl.se/docs/ipfs.html for more in"
+                      "formation\n");
+                break;
+              case CURLE_BAD_FUNCTION_ARGUMENT:
+                helpf(global->errors, "--ipfs-gateway argument results in mal"
+                      "formed URL. Visit https://curl.se/docs/ipfs.html for m"
+                      "ore information\n");
```

Comment:
please wrap the lines at word boundaries

Reply:
:)